core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -485,7 +520,39 @@ class Partition(val topic: String,
     laggingReplicas
   }
 
-  def appendRecordsToLeader(records: MemoryRecords, isFromClient: Boolean, requiredAcks: Int = 0) = {
+  def appendRecordsToFollower(records: MemoryRecords) {
+    // The read lock is needed to avoid race condition while ReplicaAlterDirThread is executing
+    // maybeDeleteAndSwapFutureReplica() to replace follower replica with the future replica
+    inReadLock(leaderIsrUpdateLock) {
+      getReplica().get.log.get.appendAsFollower(records)
+    }
+  }
+
+  def appendRecordsToFutureReplica(records: MemoryRecords) {
+    // The write lock is needed to avoid race condition while ReplicaFetcherThread is truncating log of the current replica
+    // We want to make sure if there is log truncation on the current replica, the data after that truncation point will not
+    // be appended to the future replica
+    inWriteLock(leaderIsrUpdateLock) {
```

Comment:
Yeah I have considered this option. My concern with this approach is that truncation usually happens at the Partition level or Log level, but the management of ReplicaAlterDirManager and ReplicaFetcherManager ideally should happen at the ReplicaManager level, which is higher than the level of Partition and Log. Usually the level of caller should be higher than the level of caller to avoid deadlock and to make the code logic simpler to maintain and develop in the long term.

In this specific case, if we add/remove partition from ReplicaAlterDirManager, a ReplicaFetcherThread may need to get the following lock if truncation happens: AbstractFetcherThread.partitionMapLock, Partition.leaderIsrUpdateLock, ReplicaManager.replicaStateChangeLock, ReplicaAlterDirManager.mapLock and ReplicaAlterDirThread.mapLock. This may cause deadlock if another KafkaRequestHandler attempts to get the ReplicaManager.replicaStateChangeLock and then Partition.leaderIsrUpdateLock.

Regarding your concern with "lock the writes to the current replica while writing to the future replica", I think it probably won't hurt performance much. It will reduce the maximum throughput of writing to the current replica by at most half. But I assume that for most replica, the average throughput should be much less than 50% of the maximum throughput that a broker can write to a partition. Furthermore, even if the throughput of this partition is very close to the maximum throughput, we actually want to reduce the throughput that a broker can write to the current replica before the future replica catches up. Does this make sense?

Do you think the approach I suggested previously using leader epoch would work? Do you have concern with having this approach depend on the leader epoch and thus the message format?

Reply:
@lindong28 : Your suggested approach probably works in the common case. My concerns are the following. (1) Log truncation is a relatively rare event. Your approach tries to solve the problem by requiring writes to future log to hold on to the lock to the current log on every write. So, we add overhead in the common path to solve a rare event. Intuitively, to solve a rare event, we also want to add overhead in the rare path. (2) Your approach won't be able to distinguish leader epoch inconsistency due to log truncation or bugs in Kafka. In the latter case, we probably want to error out instead of continuing. (3) The dealing with leader epoch is tricky to reason about and is already done during replica initialization in the ReplicaAlterDirThread. If possible, it would be useful to limit the places that we deal with leader epoch directly. (4) In theory, even if a topic has the new message format, a truncation may bring the log to the point where there is only old message. (5) Intuitively, the disk balancing feature seems independent of message format. So, it feels a bit weird to require that. 

As for the concerns that you mentioned on the suggested approach. I am not sure we need to hold Partition.leaderIsrUpdateLock. That lock is only needed when the replica set changes. Truncating an existing future replica doesn't change replica set though. We do want to make sure that the reinitialization in the future replica is synchronized properly with potential concurrent AlterLogDir requests, as you pointed out. So, the following is a slightly modified approach. After the ReplicaFetchThread truncates the log, it will do the following: (1) call a new method in ReplicaAlterDirThread.addPartitionsWithTruncation that takes a partition and a truncating offset. This method won't directly do truncation in the future replica. It simply sets the partition state to be truncating if the partition is still present in the partition map. When ReplicaAlterDirThread sees this new state, it ignores any pending fetch response and truncates the future log to the truncating offset, and then resumes fetching.  addPartitionsWithTruncation() needs to hold partitionMapLock. However, since only ReplicaFetchThread.truncation can call ReplicaAlterDirThread.addPartitionsWithTruncation, but not vice versa. There is no cycle to form a deadlock. This approach seems to address all the above concerns: (1) no additional overhead in the write path, which is common, (2), (3) no additional logic for dealing with leader epoch, (4) and (5) doesn't require new message format.

Neither approach deals with the case when a truncation only happens on the current replica, but not the future replica, and the broker dies. This can still lead to the case when the future replica's log can be ahead of the current replica. If the message format doesn't have the leader epoch, this may not be dealt with cleanly during future replica initialization. We could potentially just truncate the future replica to the log end offset of the current replica in that case during recovery.

Reply:
@junrao I think your points 1-5 makes sense. 

My concern with the modified approach is that it requires ReplicaFetcherThread to depend on the ReplicaAlterDirThread in order to call fetching. addPartitionsWithTruncation(). It seems a bit clumsy. I have another way of doing that by passing the truncation signal from ReplicaFetcherThread to ReplicaAlterDirThread via the affected Partition. Here is my modified approach:

1) Partition.truncateTo() and Partition.truncateFullyAndStartAt() will grab write lock of leaderIsrUpdateLock. And Partition.appendRecordsToFutureReplica() will grab read lock of leaderIsrUpdateLock. Thus we still ensure that the truncation of the current replica and the append operation of the future replica will be executed in order.
2) When the current replica of a partition is truncated, it will set Partition.futureReplicaNeedTruncation to true.
3) Partition.appendRecordsToFutureReplica() will first check whether Partition.futureReplicaNeedTruncation is true. If the flag is true, the method will set the flag to false and throw OffsetOutOfRangeException.
4) If AbstractFetcherThread.processFetchRequest() catches an OOOR exception from processPartitionData(), it will update partitionState for this partition to set truncatingLog to true. Then the future replica will be truncated based on the leader epoch later.

This approach also seems to address all the above concerns: (1) no additional overhead in the write path, which is common, (2), (3) no additional logic for dealing with leader epoch, (4) and (5) doesn't require new message format. In comparison to the previous modified approach, the class dependency graph will be simpler and more intuitive.

What do you think?

Reply:
@lindong28 : The modified approach sounds good overall. A few more questions on this. (a) Do we need the leaderIsrUpdateLock in step 1? If we truncate the current replica first and then insert the truncation point in the future replica, eventually the future replica will be able to truncate to the right offset, right? (b) For step 2 and 3, we can set the truncation state through Partition. But we could also just directly set the state in ReplicaAlterDirThread, which seems simpler since it avoids a level of indirection. (c) Not sure why we need to turn the truncation state to OOOR exception. The handling of OOOR may truncate to a different offset than the truncation offset. It seems it's clearer if we handle the truncation state directly.

Reply:
@junrao Good point regarding (a). Indeed, truncateTo() and truncateFullyAndStartAt() can simply use the readLock instead of the writeLock. readLock is needed to avoid race condition with maybeReplaceCurrentWithFutureReplica().

Regarding (b), can you explain a bit more how ReplicaFetcherThread can set state in ReplicaAlterDirThread after log truncation? Are you suggesting that ReplicaFetcherThread constructor should take `ReplicaAlterDirManger.fetcherThreadMap` as input. Then when log truncation happens for partition, it derives the corresponding ReplicaAlterDirThread from `ReplicaAlterDirManger.fetcherThreadMap` by hashing the topic and partition, before calling `ReplicaAlterThread. addPartitionsWithTruncation ()`? It seems a bit weird to put `ReplicaAlterDirManger.fetcherThreadMap` in the constructor of each ReplicaFetcherThread. What do you think?

Regarding (c), I agree we don't need to throw OOOR if we can directly set the truncation state. I am just not sure how we can directly set the truncation state due the concern with (b) described above. Also, in my suggested approach, the handling of this OOOR will not reset offset using handleOffsetOutOfRange. Instead it will set partitionState.truncatingLog to true so that the partition will be truncated based on the leader epoch later. I am going to update the patch soon so that you can see it.

Reply:
@lindong28 : I was thinking of adding a truncatePartition(Map[TopicPartition, Long]) method in AbstractFetcherThread and AbstractFetcherManager. The latter will call the former. The method takes a truncation offset for each partition in the map. ReplicaFetcherThread will have access to ReplicaAlterDirManager to call truncatePartition() when the current log is truncated. Currently, AbstractFetcherThread maintains PartitionFetchState for each partition, if the state is truncatingLog, it will issue leaderEpochRequest, followed by log truncation. We can probably introduce a separate leaderEpochState. During initialization, a partition will go through leaderEpoch state (where leader epoch request is issued) and then truncatingLog state (where log will be truncated). When truncatePartition() is called, AbstractFetcherThread just sets the partition state to truncatingLog state. When AbstractFetcherThread sees a partition in that state, it will just truncate the log to the specified truncating offset and then transition to the fetch state.

For (a), I am not sure we even need a read lock on leaderIsrUpdateLock. In the case when a partition is already removed when AbstractFetcherThread.truncatePartition() is called, we can simply ignore that partition.

Reply:
@junrao Regarding (b), it seems that leaderEpochState is not used in the suggested approach? I think the suggested approach would work. My only concern with this approach is that his approach will have ReplicaFetcherThread depend on ReplicaAlterDirThread, which is a bit unintuitive. On the other hand, my current approach is more complicated because it requires a new flag in Parition, OOOR exception from Partition.truncateTo(...), and handling of this OOOR exception thrown from processPartitionData().

I will implement the suggested approach if you think it is OK to have ReplicaFetcherThread depend on ReplicaAlterDirThread. Can you confirm this?

Regarding (a), I think readLock is needed by truncateTo() and truncateFullyAndStartAt() to avoid race condition with maybeReplaceCurrentWithFutureReplica(). For example, if truncateTo() does not grab the read lock, it is possible that maybeReplaceCurrentWithFutureReplica() finds the LEO of the future replica equals LEO of the current replica, truncateTo() truncates the current replica, and then maybeReplaceCurrentWithFutureReplica() replaces the current replica with the future replica, and the future replica has data that should have been truncated. Does this make sense?

Reply:
@lindong28 : In comparison, the approach that lets ReplicaFetcherThread reference ReplicaAlterDirManager seems a bit better.

Your explanation on the locking requirement makes sense. We are abusing the intent of leaderIsrUpdateLock now. So, we probably want to add some comments on that.

Reply:
@junrao Thanks for all the comments! I have updated the patch as suggested.